### DISCUSSION TITLE
Running a prefect server from a executable and freeze_support is not working and unable to access prefect server from localhost:4200 on browser

### DISCUSSION THREAD

**Discussion by kondasandeep13** [pyinstaller/pyinstaller#8893] [Help]
main.py

```
from multiprocessing import Process, freeze_support, set_start_method
import os
os.environ["PREFECT_API_URL"] = "http://127.0.0.1:4200/api"
os.environ["no_proxy"] = "127.0.0.1:4200"

def run_prefect_server():
    """Run the Prefect server."""
    print("Starting Prefect server...")

    try:
        server.start(host='127.0.0.1', port=4200, keep_alive_timeout=5,
                     log_level='WARNING', scheduler=True, analytics=True,
                     late_runs=True, ui=True, background=False)
    except Exception as e:
        print(f"Error starting Prefect server: {e}")

if __name__ == "__main__":

    set_start_method("spawn", force=True)
    freeze_support()
    prefect_process = Process(target=run_prefect_server)
    prefect_process.start()
    # Wait for all processes to finish
    prefect_process.join()
```

**Error**

```
Process ID: 33416
Script is executing...
Starting Prefect server...

 ___ ___ ___ ___ ___ ___ _____
| _ \ _ \ __| __| __/ __|_   _|
|  _/   / _|| _|| _| (__  | |
|_| |_|_\___|_| |___\___| |_|

Configure Prefect to communicate with the server with:

    prefect config set PREFECT_API_URL=http://127.0.0.1:4200/api

View the API reference documentation at http://127.0.0.1:4200/docs

Check out the dashboard at http://127.0.0.1:4200



16:29:31.401 | INFO    | prefect.cli.server - Opening server process with command: 'C:\Users\e40017818\Desktop\final\project\dist\main.exe' -m uvicorn --app-dir 'C:\Users\E40017~1\AppData\Local\Temp\_MEI310402' --factory prefect.server.api.server:create_app --host 127.0.0.1 --port 4200 --timeout-keep-alive 5


Process ID: 33152
Script is executing...
Starting Prefect server...

 ___ ___ ___ ___ ___ ___ _____
| _ \ _ \ __| __| __/ __|_   _|
|  _/   / _|| _|| _| (__  | |
|_| |_|_\___|_| |___\___| |_|

Configure Prefect to communicate with the server with:

    prefect config set PREFECT_API_URL=http://127.0.0.1:4200/api

View the API reference documentation at http://127.0.0.1:4200/docs

Check out the dashboard at http://127.0.0.1:4200

16:29:38.351 |

INFO    | prefect.cli.server - Opening server process with command: 'C:\Users\e40017818\Desktop\final\project\dist\main.exe' -m uvicorn --app-dir 'C:\Users\E40017~1\AppData\Local\Temp\_MEI310402' --factory prefect.server.api.server:create_app --host 127.0.0.1 --port 4200 --timeout-keep-alive 5
```
.......................
Here, I see the prefect server code is executed multiple times, and the console is having all the above lines multiple times

Help would be appreciable.

Thanks

**Comment by rokm** [comment]
Does `server.start()` spawn its own subprocess using `sys.executable`?

Try adding `print("Program arguments:", sys.argv)` at the very top of your script (and maybe also `print("Program (main) arguments:", sys.argv)` after the `freeze_support` call). This should dump arguments as passed to each instance of the program - the first print would also catch the ones from `multiprocessing`, while the second one should catch only the one from your main process (and potentially any other sub-processes that were spawned via `sys.executable` outside of `multiprocessing`).

**Reply by rokm** [reply]
Ah, nevermind, here's the answer:

```
Opening server process with command: 'C:\Users\e40017818\Desktop\final\project\dist\main.exe' -m uvicorn --app-dir 'C:\Users\E40017~1\AppData\Local\Temp\_MEI310402' --factory prefect.server.api.server:create_app --host 127.0.0.1 --port 4200 --timeout-keep-alive 5
```

So `server.start()` does spawn another instance using `sys.executable`, which then executes your main program code again.
(If this were unfrozen script, `sys.executable` would be your python interpreter executable, which would then run the `uvicorn` module's `__main__` script with provided arguments. But in the frozen application, `sys.executable` is your program executable).

You need to catch those specific command-line arguments (probably `if sys.argv[1:3] == ['-m', 'uvicorn']` should suffice), and redirect the program flow into separate function. There, import `uvicorn`, and run `uvicorn.main()` (you'll need to look into how `uvicorn.__main__` would deal with the rest of the passed command-line arguments, though).

(That's similar to what our multiprocessing run-time hook does with `freeze_support`...)
